enda/backtesting.py

```diff
@@ -1,130 +1,339 @@
+        initial_train_index = [_ for _ in range(shift_index)]
+        for train_index, test_index in time_series_split.split(sorted_df.iloc[shift_index:]):
+            train_index = initial_train_index + [_ + shift_index for _ in train_index]
+            test_index = [_ + shift_index for _ in test_index]
+            yield sorted_df.iloc[train_index], sorted_df.iloc[test_index]
+
+    @staticmethod
+    def yield_train_test_periodic_split(
+            df: pd.DataFrame,
+            test_size: Union[str, pd.Timedelta],
+            gap_size: Optional[Union[str, pd.Timedelta]] = '0D',
+            min_train_size: Optional[Union[str, pd.Timedelta]] = None
+    ) -> Generator[(pd.DataFrame, pd.DataFrame)]:
+        """
+        Returns pairs of (train set, test set) to perform back-testing on the data
+        :param df: the dataset, a pd.DataFrame with a pd.DatetimeIndex or MultiIndex with last index being
+            the DatetimeIndex.
+        :param test_size: eg. '28D', '2M'... Size of the test sample, which is the same
+            as the initial train sample, and then, size of the increase of the train sample for each iteration.
+        :param gap_size: eg '1D', '1W'... Size of the data between the train and test samples.
+        :param min_train_size: the size of the first initial train test. I None, it defaults to test_size.
+        :return: a generator of (train set, test set) pairs
+        """
+        sorted_df = df.copy().sort_index(level=-1)
+
+        # hardcode a minimal size of the test set to avoid side effects in the backtesting
+        min_test_set_size = str(TimeSeries.freq_as_approximate_nb_days(test_size) / 2) + 'D'
+        min_index = df.index.get_level_values(-1).min()
+        max_index = df.index.get_level_values(-1).max()
+
+        # initial train size
+        if min_train_size is None:
+            min_train_size = test_size
+
+        # yield train-test; each iteration adds test_size (for the new test_set) and gap_size
+        start_train = min_index
+        excl_end_train = TimeSeries.add_timedelta(start_train, min_train_size)
+        start_test = TimeSeries.add_timedelta(excl_end_train, gap_size)
+        excl_end_test = TimeSeries.add_timedelta(start_test, test_size)
+
+        while start_test < max_index:
+
+            # check test set is big enough: it's been set to half of the test_size_freq
+            # if not, it's the last run, and test set is too small to be retrieved
+            if (max_index - start_test) < pd.to_timedelta(min_test_set_size):
+                break
+
+            yield (
+                sorted_df.loc[sorted_df.index.get_level_values(-1) < excl_end_train],
+                sorted_df.loc[
+                    (sorted_df.index.get_level_values(-1) >= start_test) &
+                    (sorted_df.index.get_level_values(-1) < excl_end_test)
+                    ]
+            )
+
+            # prepare next iteration
+            excl_end_train = excl_end_test
```

Comment:
Ca engendre un comportement un peu innatendu si le `gap_size` est non nul et d'ailleurs  ca se voit dans le notebook test_backfill : 
Si on a des données journalières sur 2003 par exemple, 1 mois de test_size et 1 mois de gap, la première paire va train sur janvier et test sur février, mais du coup la deuxième paire va train sur mars parce que le nouveau train_set prend la fin du précédent test_set comme référence.
A priori, il faudrait plutôt juste ajouter le `test_size` au `excl_end_train` actuel pour obtenir le suivant ?

Reply:
Hmm, j'ai l'impression que ça fait plutôt : train sur janvier en entier, test sur mars en entier, puis train sur janvier-mars en entier, et test sur mai en entier ? 
Mais en y réfléchissant, c'est vrai qu'on zappe la période train sur janvier-février, et test sur avril. Je vais corriger ça !

Reply:
Oui c'est ca la deuxième itération est sur janvier-mars mais on skip l'entrainement sur janvier-février + test sur avril, donc par exemple pour l'algo de prev conso on serait pas dans les conditions exactes de l'algo en prod j'ai l'impression